app/services/cloud_tenant_dashboard_service.rb

```diff
@@ -47,7 +45,69 @@ def attributes_data
     format_data('cloud_tenant', attributes, attr_icon, attr_url, attr_hsh)
   end
 
-  def get_url(resource, _ems_id, attr_url)
-    "/#{resource}/show/#{@tenant_id}?display=#{attr_url}"
+  def get_url(resource, _record_id, attr_url)
+    "/#{resource}/show/#{@record_id}?display=#{attr_url}"
+  end
+
+  def get_icon(tenant)
+    fileicon = tenant.ext_management_system.decorate.try(:fileicon)
+    fileicon ? ActionController::Base.helpers.image_path(fileicon) : nil
+  end
+
+  def recent_images_data
+    recent_resources(MiqTemplate, _('Recent Images'), _('Images'))
+  end
+
+  def recent_instances_data
+    recent_resources(ManageIQ::Providers::CloudManager::Vm, _('Recent Instances'), _('Instances'))
+  end
+
+  def aggregate_status_data
+    {
+      :aggStatus => aggregate_status
+    }.compact
+  end
+
+  def aggregate_status
+    {
+      :status   => status_data,
+      :attrData => attributes_data,
+    }
+  end
+
+  def recent_resources(model, title, label)
+    all_resources = recent_records(model)
+    config = {
+      :title => title,
+      :label => label
+    }
+
+    data = if all_resources.blank?
+             {
+               :dataAvailable => false,
+               :config        => config
+             }
+           else
+             {
+               :dataAvailable => true,
+               :xData         => all_resources.keys,
+               :yData         => all_resources.values.map,
+               :config        => config
+             }
+           end
+    {
+      :recentResources => data
+    }.compact
+  end
+
+  def recent_records(model)
+    db_table     = model.arel_table
+    to_char_args = [db_table[:created_on], Arel::Nodes::SqlLiteral.new("'YYYY-MM-DD'")]
+    group_by_sql = Arel::Nodes::NamedFunction.new("to_char", to_char_args)
+
+    model.where(:ems_id => @record.ext_management_system.id)
```

Comment:
I wonder if this is right after all.

I mean, we're showing a list of vms or images whcih have the same ems as the cloud tenant.
This is definitely better than before, where we would show vms and images of a random EMS (that happened to have the same db id as the cloud tenant).

But.. don't we have a ..more directl link between cloud tenant and instances/images?

Reply:
... we do, `@record.vms` and `@record.miq_templates`.

So, what about adding a `relation` argument to `recent_records` and `recent_resources` and doing..

```ruby
def recent_records(model, relation)
    db_table     = model.arel_table
    to_char_args = [db_table[:created_on], Arel::Nodes::SqlLiteral.new("'YYYY-MM-DD'")]
    group_by_sql = Arel::Nodes::NamedFunction.new("to_char", to_char_args)

   relation.where(db_table[:created_on].gt(30.days.ago.utc))
         .group(group_by_sql.to_sql)
         .count
end

def recent_images_data
  recent_resources(MiqTemplate, _('Recent Images'), _('Images'), @record.miq_templates)
end

def recent_instances_data
  recent_resources(ManageIQ::Providers::CloudManager::Vm, _('Recent Instances'), _('Instances'), @record.vms)
end
```